// use-timer/lib/useTimer.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
require("jsdom-global/register");
var enzyme_1 = __importDefault(require("enzyme"));
var react_1 = __importDefault(require("react"));
var useTimer_1 = require("./useTimer");
var test_utils_1 = require("react-dom/test-utils");
jest.useFakeTimers();
describe('Start', function () {
    it('should start timer', function () {
        // Given
        var Component = function () {
            var _a = useTimer_1.useTimer(), time = _a.time, start = _a.start;
            return (react_1.default.createElement("div", null,
                react_1.default.createElement("button", { onClick: start }, "Start"),
                react_1.default.createElement("p", null, time)));
        };
        var wrapper = enzyme_1.default.mount(react_1.default.createElement(Component, null));
        var button = wrapper.find('button');
        var time = wrapper.find('p');
        // When
        button.simulate('click');
        test_utils_1.act(function () {
            jest.advanceTimersByTime(5000);
        });
        // Then
        expect(time.text()).toBe('5');
    });
    it('should start timer with an initial time of 10', function () {
        // Given
        var Component = function () {
            var _a = useTimer_1.useTimer({
                initialTime: 10,
            }), time = _a.time, start = _a.start;
            return (react_1.default.createElement("div", null,
                react_1.default.createElement("button", { onClick: start }, "Start"),
                react_1.default.createElement("p", null, time)));
        };
        var wrapper = enzyme_1.default.mount(react_1.default.createElement(Component, null));
        var button = wrapper.find('button');
        var time = wrapper.find('p');
        // When
        button.simulate('click');
        test_utils_1.act(function () {
            jest.advanceTimersByTime(5000);
        });
        // Then
        expect(time.text()).toBe('15');
    });
    it('should start decremental timer with an initial time of 100', function () {
        // Given
        var Component = function () {
            var _a = useTimer_1.useTimer({
                initialTime: 100,
                timerType: 'DECREMENTAL',
            }), time = _a.time, start = _a.start;
            return (react_1.default.createElement("div", null,
                react_1.default.createElement("button", { onClick: start }, "Start"),
                react_1.default.createElement("p", null, time)));
        };
        var wrapper = enzyme_1.default.mount(react_1.default.createElement(Component, null));
        var button = wrapper.find('button');
        var time = wrapper.find('p');
        // When
        button.simulate('click');
        test_utils_1.act(function () {
            jest.advanceTimersByTime(20000);
        });
        // Then
        expect(time.text()).toBe('80');
    });
    it('should update time with an interval of 2000 milliseconds', function () {
        // Given
        var Component = function () {
            var _a = useTimer_1.useTimer({
                interval: 2000,
            }), time = _a.time, start = _a.start;
            return (react_1.default.createElement("div", null,
                react_1.default.createElement("button", { onClick: start }, "Start"),
                react_1.default.createElement("p", null, time)));
        };
        var wrapper = enzyme_1.default.mount(react_1.default.createElement(Component, null));
        var button = wrapper.find('button');
        var time = wrapper.find('p');
        // When
        button.simulate('click');
        test_utils_1.act(function () {
            jest.advanceTimersByTime(10000);
        });
        // Then
        expect(time.text()).toBe('5');
    });
});
describe('Stop', function () {
    it('should stop incremental timer when time is over', function () {
        // Given
        var Component = function () {
            var _a = useTimer_1.useTimer({
                endTime: 25,
                initialTime: 5,
            }), time = _a.time, start = _a.start;
            return (react_1.default.createElement("div", null,
                react_1.default.createElement("button", { onClick: start }, "Start"),
                react_1.default.createElement("p", null, time)));
        };
        var wrapper = enzyme_1.default.mount(react_1.default.createElement(Component, null));
        var button = wrapper.find('button');
        var time = wrapper.find('p');
        // When
        button.simulate('click');
        test_utils_1.act(function () {
            jest.advanceTimersByTime(40000);
        });
        // Then
        expect(time.text()).toBe('25');
    });
    it('should stop decremental timer when time is over', function () {
        // Given
        var Component = function () {
            var _a = useTimer_1.useTimer({
                endTime: 10,
                initialTime: 30,
                timerType: 'DECREMENTAL',
            }), time = _a.time, start = _a.start;
            return (react_1.default.createElement("div", null,
                react_1.default.createElement("button", { onClick: start }, "Start"),
                react_1.default.createElement("p", null, time)));
        };
        var wrapper = enzyme_1.default.mount(react_1.default.createElement(Component, null));
        var button = wrapper.find('button');
        var time = wrapper.find('p');
        // When
        button.simulate('click');
        test_utils_1.act(function () {
            jest.advanceTimersByTime(30000);
        });
        // Then
        expect(time.text()).toBe('10');
    });
});
describe('Pause', function () {
    it('should pause timer', function () {
        // Given
        var Component = function () {
            var _a = useTimer_1.useTimer(), time = _a.time, start = _a.start, pause = _a.pause;
            return (react_1.default.createElement("div", null,
                react_1.default.createElement("button", { id: "start", onClick: start }, "Start"),
                react_1.default.createElement("button", { id: "pause", onClick: pause }, "Start"),
                react_1.default.createElement("p", null, time)));
        };
        var wrapper = enzyme_1.default.mount(react_1.default.createElement(Component, null));
        var startButton = wrapper.find('#start');
        var pauseButton = wrapper.find('#pause');
        var time = wrapper.find('p');
        // When
        startButton.simulate('click');
        test_utils_1.act(function () {
            jest.advanceTimersByTime(5000);
        });
        pauseButton.simulate('click');
        test_utils_1.act(function () {
            jest.advanceTimersByTime(5000);
        });
        // Then
        expect(time.text()).toBe('5');
    });
    it('should pause timer with an end time', function () {
        // Given
        var Component = function () {
            var _a = useTimer_1.useTimer({
                endTime: 5,
            }), time = _a.time, start = _a.start, pause = _a.pause;
            return (react_1.default.createElement("div", null,
                react_1.default.createElement("button", { id: "start", onClick: start }, "Start"),
                react_1.default.createElement("button", { id: "pause", onClick: pause }, "Start"),
                react_1.default.createElement("p", null, time)));
        };
        var wrapper = enzyme_1.default.mount(react_1.default.createElement(Component, null));
        var startButton = wrapper.find('#start');
        var pauseButton = wrapper.find('#pause');
        var time = wrapper.find('p');
        // When
        startButton.simulate('click');
        test_utils_1.act(function () {
            jest.advanceTimersByTime(3000);
        });
        pauseButton.simulate('click');
        test_utils_1.act(function () {
            jest.advanceTimersByTime(5000);
        });
        startButton.simulate('click');
        // Then
        expect(time.text()).toBe('3');
        test_utils_1.act(function () {
            jest.advanceTimersByTime(3000);
        });
        expect(time.text()).toBe('5');
    });
});
describe('Reset', function () {
    it('should reset timer to default initial time', function () {
        // Given
        var Component = function () {
            var _a = useTimer_1.useTimer(), time = _a.time, start = _a.start, reset = _a.reset;
            return (react_1.default.createElement("div", null,
                react_1.default.createElement("button", { id: "start", onClick: start }, "Start"),
                react_1.default.createElement("button", { id: "reset", onClick: reset }, "Reset"),
                react_1.default.createElement("p", null, time)));
        };
        var wrapper = enzyme_1.default.mount(react_1.default.createElement(Component, null));
        var startButton = wrapper.find('#start');
        var resetButton = wrapper.find('#reset');
        var time = wrapper.find('p');
        // When
        startButton.simulate('click');
        test_utils_1.act(function () {
            jest.advanceTimersByTime(5000);
        });
        resetButton.simulate('click');
        // Then
        expect(time.text()).toBe('0');
    });
    it('should reset timer to default initial time after restart', function () {
        // Given
        var Component = function () {
            var _a = useTimer_1.useTimer({
                endTime: 10,
            }), time = _a.time, start = _a.start;
            return (react_1.default.createElement("div", null,
                react_1.default.createElement("button", { id: "start", onClick: start }, "Start"),
                react_1.default.createElement("p", null, time)));
        };
        var wrapper = enzyme_1.default.mount(react_1.default.createElement(Component, null));
        var startButton = wrapper.find('#start');
        var time = wrapper.find('p');
        // When
        startButton.simulate('click');
        test_utils_1.act(function () {
            jest.advanceTimersByTime(10000);
        });
        startButton.simulate('click');
        test_utils_1.act(function () {
            jest.advanceTimersByTime(5000);
        });
        // Then
        expect(time.text()).toBe('5');
    });
    it('should reset timer to initial time of 20', function () {
        // Given
        var Component = function () {
            var _a = useTimer_1.useTimer({
                initialTime: 20,
            }), time = _a.time, start = _a.start, reset = _a.reset;
            return (react_1.default.createElement("div", null,
                react_1.default.createElement("button", { id: "start", onClick: start }, "Start"),
                react_1.default.createElement("button", { id: "reset", onClick: reset }, "Reset"),
                react_1.default.createElement("p", null, time)));
        };
        var wrapper = enzyme_1.default.mount(react_1.default.createElement(Component, null));
        var startButton = wrapper.find('#start');
        var resetButton = wrapper.find('#reset');
        var time = wrapper.find('p');
        // When
        startButton.simulate('click');
        test_utils_1.act(function () {
            jest.advanceTimersByTime(5000);
        });
        resetButton.simulate('click');
        // Then
        expect(time.text()).toBe('20');
    });
});
describe('State and callbacks', function () {
    it('should display "Running" text when timer is running', function () {
        // Given
        var Component = function () {
            var _a = useTimer_1.useTimer({
                initialTime: 20,
            }), isRunning = _a.isRunning, start = _a.start, pause = _a.pause, reset = _a.reset;
            return (react_1.default.createElement("div", null,
                react_1.default.createElement("button", { id: "start", onClick: start }, "Start"),
                react_1.default.createElement("button", { id: "pause", onClick: pause }, "Pause"),
                react_1.default.createElement("button", { id: "reset", onClick: reset }, "Reset"),
                react_1.default.createElement("p", null, isRunning ? 'Running' : 'Not running')));
        };
        var wrapper = enzyme_1.default.mount(react_1.default.createElement(Component, null));
        var startButton = wrapper.find('#start');
        var pauseButton = wrapper.find('#pause');
        var resetButton = wrapper.find('#reset');
        var statusBlock = wrapper.find('p');
        // When
        startButton.simulate('click');
        // Then
        expect(statusBlock.text()).toBe('Running');
        // When
        pauseButton.simulate('click');
        // Expect
        expect(statusBlock.text()).toBe('Not running');
        // When
        startButton.simulate('click');
        // Expect
        expect(statusBlock.text()).toBe('Running');
        // When
        resetButton.simulate('click');
        // Expect
        expect(statusBlock.text()).toBe('Not running');
    });
    it('should call callback function when time is over', function () {
        // Given
        var onTimeOver = jest.fn();
        var Component = function () {
            var start = useTimer_1.useTimer({
                endTime: 30,
                initialTime: 0,
                onTimeOver: onTimeOver,
            }).start;
            return (react_1.default.createElement("div", null,
                react_1.default.createElement("button", { id: "start", onClick: start }, "Start")));
        };
        var wrapper = enzyme_1.default.mount(react_1.default.createElement(Component, null));
        var startButton = wrapper.find('#start');
        // When
        startButton.simulate('click');
        test_utils_1.act(function () {
            jest.advanceTimersByTime(30000);
        });
        // Then
        expect(onTimeOver).toHaveBeenCalled();
    });
});
